### PR TITLE
PB-159: remove weights from gRPC messages

### PR DESCRIPTION
**Pull request by little-dude**
## References:

https://xainag.atlassian.net/browse/PB-159

Needs to be merged along with:

- https://github.com/xainag/xain-proto/pull/25
- https://github.com/xainag/xain-sdk/pull/88
- https://github.com/xainag/xain-fl/pull/298

## Summary:

Remove the weights from the gRPC messages. From now on, weights will
be exchanged via s3 buckets.

The sequence diagram below illustrate this new behavior.

At the beginning of a round (1) the selected participants send a
`StartTrainingRound` request, and the coordinator response with the
same `StartTrainingRoundResponse` that does not contain the global
weights anymore.

Instead, the participant fetches these weights from the store (2). S3
buckets are key-value stores, and the key for global weights is the
round number.

Then, the participant trains. Once done, it uploads its local weights
to the S3 bucket (3). The key is `<round_number>/<participant_id>`.

Finally (4), the participant sends it's `EndTrainingRequest`. Before
answering, the coordinator retrieves the local weights the participant
has uploaded.

_**Important note**: At the moment, the participants don't know their
ID, because the coordinator does not send it to them. Thus, they
currently generate a random ID when they start, and send it to the
coordinator so that it can retrieve the participant's weights. This is
why the `EndTrainingRoundRequest` currently has a `participant_id`
field._

```
    P                                C                      Store
1.  |   StartTrainingRoundRequest    |                        |
    | -----------------------------> |                        |
    |   StartTrainingRoundResponse   |                        |
    | <----------------------------- |                        |
    |                                |                        |
    |                Get global weights (key="round/global")  |
2.  | ------------------------------------------------------> |
    |                         Global weights                  |
    | <------------------------------------------------------ |
    |                                |                        |
    | [train...]                     |                        |
    |                                |                        |
3.  |       Set local weights (key="round/participant")       |
    | ------------------------------------------------------> |
    |                               Ok                        |
    | <------------------------------------------------------ |
    |                                |                        |
4.  |   EndTrainingRoundRequest      |                        |
    | -----------------------------> | Get local weights (key="round/participant")
    |                                | ---------------------> |
    |                                | Local weights          |
    |  EndTrainingRoundResponse      | <--------------------> |
    | <----------------------------- |                        |
```

At the end of the round, the coordinator writes the weights to the s3
bucket, using the next upcoming round number as key (see the sequence
diagram below).

```
P                                C                      Store
|   EndTrainingRoundRequest      |                        |
| -----------------------------> | Get local weights (key="round/participant")
|                                | ---------------------> |
|                                | Local weights          |
|  EndTrainingRoundResponse      | <--------------------> |
| <----------------------------- |                        |
|                                |                        |
|                                | Set global weights (key="round+1/participant")
|                                | ---------------------> |
|                                | Ok                     |
|                                | <--------------------> |
```

Implementation notes:

- Initially, we thought we would be using different buckets for the
  local and global weights. But for now, we use the same bucket for
  local and global weights for now

- We currently store the global weights under different keys. It turns
  out that this brings un-necessary complexity so we'll probably
  simplify this in the future

- For now, the coordinator doesn't send any storage information to the
  participants. Thus, the participants need to be configured with the
  storage information. In the future, the `StartTrainingRoundResponse`
  could contain the endpoint url, bucket name, etc.

---

### Reviewer checklist

*Reviewer agreement:*

* Reviewers assign themselves at the start of the review.
* Reviewers do **not** commit or merge the merge request.
* Reviewers have to check and mark items in the checklist.

**Merge request checklist**

- [x] Conforms to the merge request title naming `XP-XXX <a description in imperative form>`.
- [x] Each commit conforms to the naming convention `XP-XXX <a description in imperative form>`.
- [x] Linked the ticket in the merge request title or the references section.
- [x] Added an informative merge request summary.

**Code checklist**

- [x] Conforms to the branch naming `XP-XXX-<a_small_stub>`.
- [x] Passed scope checks.
- [x] Added or updated tests if needed.
- [x] Added or updated code documentation if needed.
- [x] Conforms to Google docstring style.
- [x] Conforms to XAIN structlog style.

### CODE REVIEW

File: protobuf/xain_proto/fl/coordinator.proto [hunk 1]
```diff
@@ -48,13 +48,12 @@ message HeartbeatResponse {
 message StartTrainingRoundRequest {}
 
 message StartTrainingRoundResponse {
-  xain_proto.np.NDArray weights = 1; // model weights
   int32 epochs = 2;                  // number of epochs for the training round
   int32 epoch_base = 3;              // global base epoch
 }
```
Comment (janpetschexain) [inline protobuf/xain_proto/fl/coordinator.proto]: i know that one usually shouldn't change the field numbers in protobuf messages for backward compatibility, but since we are the only guys using this at the moment, it's probably ok to reorder that to
```
message StartTrainingRoundResponse {
  int32 epochs = 1;     // number of epochs for the training round
  int32 epoch_base = 2; // global base epoch
}
```

Comment (little-dude) [reply protobuf/xain_proto/fl/coordinator.proto]: Ah thanks, I wasn't sure about this. Will do.

Comment (finiteprods) [reply protobuf/xain_proto/fl/coordinator.proto]: on the other extreme, for backwards-compatibility should we even remove this? I don't think there is harm in keeping this (e.g. for testing purposes). Maybe we could just comment that it is "deprecated, in favour of..."

**Comment by little-dude** [comment]
> I wonder whether we need to completely retire the option of transmitting weights over the wire. After all, it does work (with some niggles), so it may be useful to keep the option whether as a fall-back or for testing.

I partialy answered this in the `xain-sdk` PR ([here](https://github.com/xainag/xain-sdk/pull/88#discussion_r380584238)):

> It is not easy to make storage optional unfortunately.If we don't use storage for the weights, we need to exchange them via the gRPC messages. This is a completely different distribution model, so it would probably require specific logic in multiple places, in both the coordinator and sdk.

I see two issues with keeping both weights distribution mechanisms:

- it would make code more complex (we have to check which distribution mechanism is used, check for corner cases in both case, etc)
- it would be a maintenance burden (we'd have to add tests for both mechanisms in some cases, ensure that we don't break one when touching the other, etc)

File: protobuf/xain_proto/fl/coordinator.proto (hunk 1 above)
Comment (little-dude) [reply protobuf/xain_proto/fl/coordinator.proto]: I have no opinion here. Backwards-compatibility is not really an issue for us since we haven't had a stable release yet. But as you say it doesn't harm. @atymoshchuk @Robert-Steiner do you have a preference?

Comment (Robert-Steiner) [reply protobuf/xain_proto/fl/coordinator.proto]: I agree with @little-dude. Since it is still in alpha, we should try to keep the code simple and clean.
